pkg/util/grpcutil.go

```diff
@@ -0,0 +1,221 @@
+	}
+
+	certificate, err := tls.X509KeyPair(cert, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return serverConfig(certPool, &certificate, peerName), nil
+}
+
+// LoadServerTLS prepares the TLS configuration needed for a server with the given certificate files.
+// peerName is either the name that the client is expected to have a certificate for or empty,
+// in which case any client is allowed to connect.
+func LoadServerTLS(caFile, certFile, keyFile, peerName string) (*tls.Config, error) {
+	certPool, peerCert, err := loadCertificate(caFile, certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+	return serverConfig(certPool, peerCert, peerName), nil
+}
+
+func serverConfig(certPool *x509.CertPool, peerCert *tls.Certificate, peerName string) *tls.Config {
+	return &tls.Config{
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (*tls.Config, error) {
+			if info == nil {
+				return nil, errors.New("nil client info passed")
+			}
+
+			config := &tls.Config{
+				MinVersion:    tls.VersionTLS13,
+				Renegotiation: tls.RenegotiateNever,
+				Certificates:  []tls.Certificate{*peerCert},
+				ClientCAs:     certPool,
+				VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+					// Common name check when accepting a connection from a client.
+					if peerName == "" {
+						// All names allowed.
+						return nil
+					}
+
+					if len(verifiedChains) == 0 ||
+						len(verifiedChains[0]) == 0 {
+						return errors.New("no valid certificate")
+					}
+
+					for _, name := range verifiedChains[0][0].DNSNames {
+						if name == peerName {
+							return nil
+						}
+					}
+					// For compatibility - using CN as hostName
+					commonName := verifiedChains[0][0].Subject.CommonName
+					if commonName == peerName {
+						return nil
+					}
+					return fmt.Errorf("certificate is not signed for %q hostname", peerName)
+				},
+			}
+			if peerName != "" {
+				config.ClientAuth = tls.RequireAndVerifyClientCert
+			}
+			return config, nil
+		},
+	}
+}
+
+// ClientTLS prepares the TLS configuration that can be used by a client while connecting to a server
+// with given encoded certificate and private key.
+// peerName must be provided when expecting the server to offer a certificate with that CommonName.
+func ClientTLS(caCert, cert, key []byte, peerName string) (*tls.Config, error) {
+	certPool := x509.NewCertPool()
+	if ok := certPool.AppendCertsFromPEM(caCert); !ok {
+		return nil, fmt.Errorf("failed to append CA certificate to pool")
+	}
+
+	certificate, err := tls.X509KeyPair(cert, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientConfig(certPool, &certificate, peerName), nil
+}
+
+// LoadClientTLS prepares the TLS configuration that can be used by a client while connecting to a server.
+// peerName must be provided when expecting the server to offer a certificate with that CommonName. caFile, certFile, and keyFile are all optional.
```

Comment:
I believe at least need to provide `caFile`, otherwise how to verify the server connection? 

`certFile` and `keyFile` are needed only when we would like to have mTLS to let sever/client identify each other. I believe we will introduce below cases:
- longhorn-manager -> longhorn-instance-manager/engine via TLS connect not mutual
- longhorn-engine -> longhorn-replica via TLS only or mTLS?
- longhorn-replica <-> longhorn-replica via TLS only or mTLS? 

Probably need to check longhorn-manager PR later, but introducing TLS in 1.3.0 should be fine as the first phase.

Reply:
I am confused about this part as well. Why are these authentication/encryption-related files optional while the server name is required?

Reply:
@joshimoo any update? After resolving all comments, we can merge this.

Reply:
Let me try to clarify:
- caFile is optional since the default are the root ca's shipped by linux
   - Example case where this is important, if you have a valid certificate produced by a root CA for example one doesn't need to provide a CA file since it's already part of the root CA chain. for example [let's encrypt](https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/)
- certFile&keyfile are optional in the client case, since they are only required when client authentication is used (Mutual TLS)
   - Forwarding thinking I could envision a use case where a client wants to connect to regular TLS server without MTLS, example use case are health/metrics endpoint, etc
- peerName is optional too, but requires a little bit more SSL knowledge to understand what it's used for, I will leave some links for context / reading below
   - Since we connect against an IP address instead of a DNS name, no host name would be available so we provide a peerName to check that the server/client returned certificate is for a specific set of hosts.
   - [Practical explanation](https://github.com/golang/go/issues/7342#issuecomment-66091467) ,[RFC TLS SNI](https://datatracker.ietf.org/doc/html/rfc6066#section-3)

I generally tried to keep this simple and generic, if you feel like hard coding more of this or removing the flexibility would be preferred let me know.

Reply:
Thanks, it's good now, especially for the clarification of the optional caFile.